pkg/internalapis/featurestates/featurestates.go

```
@@ -0,0 +1,418 @@
+	supervisorFeatureStateConfigMapNamespace string
+	pendingCRUpdatesObj                      *pendingCRUpdates
+	k8sClient                                clientset.Interface
+	controllerRuntimeClient                  client.Client
+)
+
+// StartSvFSSReplicationService Starts SvFSSReplicationService
+func StartSvFSSReplicationService(ctx context.Context, svFeatureStatConfigMapName string, svFeatureStateConfigMapNamespace string) error {
+	log := logger.GetLogger(ctx)
+	log.Info("Starting SvFSSReplicationService")
+
+	supervisorFeatureStatConfigMapName = svFeatureStatConfigMapName
+	supervisorFeatureStateConfigMapNamespace = svFeatureStateConfigMapNamespace
+	pendingCRUpdatesObj = &pendingCRUpdates{
+		lock:                &sync.RWMutex{},
+		namespaceUpdateMap:  make(map[string]bool),
+		latestFeatureStates: make([]featurestatesv1alpha1.FeatureState, 0),
+	}
+	var err error
+	// This is idempotent if CRD is pre-created then we continue with initialization of svFSSReplicationService
+	err = k8s.CreateCustomResourceDefinitionFromSpec(ctx, CRDName, CRDSingular, CRDPlural,
+		reflect.TypeOf(featurestatesv1alpha1.CnsCsiSvFeatureStates{}).Name(), CRDGroupName, internalapis.SchemeGroupVersion.Version, apiextensionsv1beta1.NamespaceScoped)
+	if err != nil {
+		log.Errorf("failed to create CnsCsiSvFeatureStates CRD. Error: %v", err)
+		return err
+	}
+
+	// Create the kubernetes client
+	k8sClient, err = k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("create k8s client failed. Err: %v", err)
+		return err
+	}
+	// get kube config
+	config, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		log.Errorf("failed to get kubeconfig. Error: %v", err)
+		return err
+	}
+	// create controller runtime client
+	controllerRuntimeClient, err = k8s.NewClientForGroup(ctx, config, CRDGroupName)
+	if err != nil {
+		log.Errorf("failed to create controllerRuntimeClient. Err: %v", err)
+		return err
+	}
+
+	pendingCRUpdatesObj.latestFeatureStates, err = getFeatureStates(ctx)
```

Now that you are listening to config map events, do you need this?

---

Kept it here, to fail the `StartSvFSSReplicationService` if fss config-map is not present.

---

Why do you want to fail if config map is not present? Since you anyways have a listener and you are handling create event, it should be fine, right? I'm wondering if we are adding redundant code and logic here which is not necessary.

---

Can you respond to this comment?

---

we are starting processPendingCRUpdates() as go routine, so there is a chance that `processPendingCRUpdates()` may get started before config-map listener done executing `configMapAdded`.

I am removing it from here and adding in inside processPendingCRUpdates() to handle that case.

```golang
// processPendingCRUpdates helps process pending CR updates at regular interval
func (pendingCRUpdatesObj *pendingCRUpdates) processPendingCRUpdates() {
	ticker := time.NewTicker(crUpdateRetryInterval)
	for range ticker.C {
		func() {
			pendingCRUpdatesObj.lock.Lock()
			defer pendingCRUpdatesObj.lock.Unlock()

			ctx, log := logger.GetNewContextWithLogger()
			var err error
			if len(pendingCRUpdatesObj.latestFeatureStates) == 0 {
				log.Warn("empty feature states observed. feature state config-map might be deleted. Trying to obtain latest feature states.")
				pendingCRUpdatesObj.latestFeatureStates, err = getFeatureStates(ctx)
				if err != nil {
					log.Errorf("Failed to get feature states. error: %v", err)
					os.Exit(1)
				}
			}
.
.
.
```